Categoricals hash consistently

jcrist opened the pull request:
Previously categorical values were hashed using just their codes. This
meant that the hash value depended on the ordering of the categories,
rather than on the values the series represented. This caused problems
in dask, where different partitions might have different categorical
mappings.

This PR makes the hashing dependent on the values the categorical
represents, rather than on the codes. The categories are first hashed,
and then the codes are remapped to the hashed values. This is slightly
slower than before (still need to hash the categories, where we didn't
before), but allows for more consistent hashing.

Related to this work in dask: https://github.com/dask/dask/pull/1877.

jcrist committed:
Categoricals hash consistently

Previously categorical values were hashed using just their codes. This
meant that the hash value depended on the ordering of the categories,
rather than on the values the series represented. This caused problems
in dask, where different partitions might have different categorical
mappings.

This PR makes the hashing dependent on the values the categorical
represents, rather than on the codes. The categories are first hashed,
and then the codes are remapped to the hashed values. This is slightly
slower than before (still need to hash the categories, where we didn't
before), but allows for more consistent hashing.

jreback commented on pandas/tools/hashing.py:
```diff
@@ -80,53 +80,56 @@ def hash_array(vals, encoding='utf8', hash_key=None):
+    # numpy if categorical is a subdtype of complex, as it will choke.
     if is_categorical_dtype(vals.dtype):
-        vals = vals.codes
+        cat_hashed = hash_array(vals.categories.values, encoding, hash_key,
```
do this like below instead, rather than adding a keyword (where we do basically the same thing)

jcrist replied:
I'm not sure I follow - we need to recurse through in case the values aren't objects. So we need to be able to handle `vals.categories.values` of any dtype. However, we don't want to categorize the categories again, since we already know they're unique. This seemed to me the simplest and cleanest way of implementing this.

jreback replied:
maybe show what the problem is
hash_array can hash values JUST like we do below

this code is too complicated

jcrist replied:
Can you explain where you mean "like we do below"? I'm assuming you mean lines 125-130, where we handle object dtype.

For categoricals, what we want to do is:
- Get a hash for the `categories`. This should match what `hash_pandas_object(series, index=False)` would return for the un-categorized data. Meaning `hash_pandas_object(object_series) == hash_pandas_object(object_series.astype('category'))`.
- Remap the category hashes based on the codes.

`vals.categories.values` can be of any dtype, so we need to recurse through `hash_array` again to get the hashes of the categories. However, we also know that the categories are already unique, so we don't want to call `factorize` again. As such, we set `categorize=False` to skip that. We can't just call `hash_object_array`, as the categories may not be objects. And we need to do the remapping, so we can't just set `vals = something` and fall through like we did before.

I don't think adding this extra keyword overly complicates things, and do think this is the simplest way to do this. I may not be understanding what you're trying to suggest here - perhaps if you could explain a bit better I might get it.

jreback replied:
I looked again. This is reasonable. Ideally don't want to repeat the factorize-hash-remap logic as right now this code looks very similar to what you have. maybe pull this out to a common helper function?

I get the problem is now ``hash_array_object`` will only de-duplicate object types, when of course we could have any dtypes. Though in practice it is generally not a big deal to simply hash non-objects even if very many duplicates (as simply finding the duplicates can be somewhat expensive). In my tests, only for object does this make a huge difference.

That said not averse to allow a caller to do this (as dask may have more information that pandas, e.g. the cardinaility of the set, or even the uniques already?)

jcrist replied:
> I get the problem is now hash_array_object will only de-duplicate object types, when of course we could have any dtypes. Though in practice it is generally not a big deal to simply hash non-objects even if very many duplicates (as simply finding the duplicates can be somewhat expensive). In my tests, only for object does this make a huge difference.

Not quite. The `categorize` parameter only indicates whether to `categorize` *object arrays* before hashing. Other arrays are treated the same as before. The reason for this is that when we hash the `categories` attribute on a categorical (and then map the codes to the hashes of the categories), we already know that it is unique, so we don't need to re-categorize before hashing (which would happen for object categoricals, but not for others). This might be more clear now that the `_hash_categorical` helper is pulled out.

jreback commented on pandas/tools/hashing.py:
```diff
@@ -80,53 +80,56 @@ def hash_array(vals, encoding='utf8', hash_key=None):
+    categorize : bool, default True
+        Whether to first categorize object arrays before hashing. This is more
+        efficient when the array contains duplicate values.
 
```
add a versionadded tag here (0.20.0)

jreback replied:
I think also add ``categorize`` to hash_pandas_object for consistency as well?

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -309,6 +309,7 @@ Bug Fixes
 - Bug in ``pd.read_csv()`` in which the ``dialect`` parameter was not being verified before processing (:issue:`14898`)
 - Bug in ``pd.read_fwf`` where the skiprows parameter was not being respected during column width inference (:issue:`11256`)
 - Bug in ``pd.read_csv()`` in which missing data was being improperly handled with ``usecols`` (:issue:`6710`)
+- Bug in ``pandas.tools.hashing.hash_pandas_object`` in which hashing of categoricals depended on the ordering of categories, instead of just their values.
```
put a ``()`` as this is a function call, and ``pandas`` -> ``pd``

jreback replied:
add this PR number as the issue number

jreback commented on pandas/tools/tests/test_hashing.py:
```diff
@@ -90,6 +90,20 @@ def test_hash_pandas_empty_object(self):
+    def test_categorical_consistency(self):
+        # Check that categoricals hash consistent with their values, not codes
+        # This should work for categoricals of any dtype
+        for data in [['a', 'b', 'c', 'd'], [1000, 2000, 3000, 4000]]:
```
can you test with datetimes as well

jcrist committed:
Address comments

- Add `categorize` parameter to `hash_pandas_object`
- Update test
- Update whatsnew

jcrist commented:
Thanks for the review. Addressed all comments.

jreback commented:
lgtm.

ping on green. (note appveyor is currently recovering, so may take a while) :<

jcrist commented:
Windows tests passed, travis failures look unrelated?

jreback commented:
thanks @jcrist
